src/ui/public/indexed_array/index.d.ts

```diff
@@ -0,0 +1,41 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { ListIterator } from 'lodash';
+
+interface IndexedArrayConfig<T> {
+  index?: string[];
+  group?: string[];
+  order?: string[];
+  initialSet?: T[];
+  immutable?: boolean;
+}
+
+declare class IndexedArray<T> extends Array<T> {
+  public immutable: boolean;
+  public raw: T[];
+  // May not actually be present, is dynamically defined.
+  public inOrder: T[];
+
+  constructor(config: IndexedArrayConfig<T>);
+
+  public remove(predicate: ListIterator<T, boolean>): T[];
+
+  public toJSON(): T[];
+}
```

Comment:
This whole type definition is pretty bad right now

Reply:
Suggestion, after I already looked into trying to type this. TypeScript can't express that dynamic method names we have here. Instead of adding `inOrder` etc manually here, since they might anyway not exist, we could potentially just give this class besides the typed methods the index signature `[key: string]: T[]`. so it would automatically assume every non existing method is that kind of array?